pkg/net/libp2p/libp2p_transport_test.go

```diff
@@ -0,0 +1,309 @@
+
+	t.Log(">>> connecting nodes")
+
+	initialNodes := m.Membership(nodeA, nodeB, nodeC, nodeD)
+	initialNodes[nodeE] = eAddr
+
+	t.Log("membership")
+	t.Log(initialNodes)
+
+	a.Connect(ctx, initialNodes)
+	b.Connect(ctx, initialNodes)
+	c.Connect(ctx, initialNodes)
+	d.Connect(ctx, initialNodes)
+
+	m.testEventuallyConnected(nodeA, nodeB)
+	m.testEventuallyConnected(nodeA, nodeC)
+	m.testEventuallyConnected(nodeA, nodeD)
+	m.testEventuallyConnected(nodeB, nodeC)
+	m.testEventuallyConnected(nodeB, nodeD)
+	m.testEventuallyConnected(nodeC, nodeD)
+
+	t.Log(">>> sending messages")
+	var nodeErr *nodeInfoError
+	err = a.Send(ctx, "unknownNode", &messagepb.Message{})
+	require.ErrorAs(t, err, &nodeErr)
+
+	var streamErr *nilStreamError
+	err = a.Send(ctx, nodeE, &messagepb.Message{})
+	require.ErrorAs(t, err, &streamErr)
+
+	nodeBEventsChan := b.EventsOut()
+	nodeCEventsChan := c.EventsOut()
+
+	err = a.Send(ctx, nodeB, &messagepb.Message{})
+	require.NoError(t, err)
+	nodeBEvents := <-nodeBEventsChan
+	msg, valid := nodeBEvents.Iterator().Next().Type.(*eventpb.Event_MessageReceived)
+	require.Equal(t, true, valid)
+	require.Equal(t, msg.MessageReceived.From, nodeA.Pb())
+
+	err = a.Send(ctx, nodeC, &messagepb.Message{})
+	require.NoError(t, err)
+	nodeCEvents := <-nodeCEventsChan
+	msg, valid = nodeCEvents.Iterator().Next().Type.(*eventpb.Event_MessageReceived)
+	require.Equal(t, true, valid)
+	require.Equal(t, msg.MessageReceived.From, nodeA.Pb())
+
+	t.Log(">>> disconnecting nodes")
+	m.disconnect(nodeA, nodeB)
+	m.testEventuallyNotConnected(nodeA, nodeB)
+	require.Equal(t, 2, len(b.host.Network().Peers()))
+
+	t.Log(">>> sending messages after disconnection")
+	m.testEventuallySentMsg(ctx, nodeA, nodeB, &messagepb.Message{})
+
+	nodeBEvents = <-nodeBEventsChan
+	msg, valid = nodeBEvents.Iterator().Next().Type.(*eventpb.Event_MessageReceived)
+	require.Equal(t, true, valid)
+	require.Equal(t, msg.MessageReceived.From, nodeA.Pb())
+
+	m.testEventuallySentMsg(ctx, nodeA, nodeC, &messagepb.Message{})
+	nodeCEvents = <-nodeCEventsChan
+	msg, valid = nodeCEvents.Iterator().Next().Type.(*eventpb.Event_MessageReceived)
+	require.Equal(t, true, valid)
+	require.Equal(t, msg.MessageReceived.From, nodeA.Pb())
+}
+
+func TestLibp2p_Connecting(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	logger := logging.ConsoleDebugLogger
+
+	nodeA := types.NodeID("a")
+	nodeB := types.NodeID("b")
+	nodeC := types.NodeID("c")
+	nodeD := types.NodeID("d")
+
+	m := newMockLibp2pCommunication(t, []types.NodeID{nodeA, nodeB, nodeC, nodeD}, logger)
+
+	a, b, c, d := m.FourTransports(nodeA, nodeB, nodeC, nodeD)
+	m.StartAll()
+	defer m.StopAll()
+
+	t.Log(">>> connecting nodes")
+	t.Log("membership")
+	t.Log(m.membership)
+
+	initialNodes := m.Membership(nodeA, nodeB, nodeC)
+
+	a.Connect(ctx, initialNodes)
+	b.Connect(ctx, initialNodes)
+	c.Connect(ctx, initialNodes)
+
+	m.testEventuallyConnected(nodeA, nodeB)
```

Comment:
Should we also test if the retries work and have some parallel tests (with calls from different routines)? I haven´t looked in depth into the code, so maybe they are already there.

Reply:
(This applies to some of the previous comments too.)

In the current implementation, there is no way a connection could be written to concurrently by multiple threads. The whole module is handled by a single thread, sending messages one by one. This is also why getting stuck on writing to one broken (or slow) connection is such a big problem. However, at least we don't have to be concerned about locking the streams when writing into them (unless we need to synchronize the writes with potential reconnections).

Reply:
We have a basic check that retry works - https://github.com/filecoin-project/mir/blob/be26a6d97e5d97d0c2a87ba4056f5d16cfc0a58f/pkg/net/libp2p/libp2p_transport_test.go#L235